cmd/kubeadm/app/phases/upgrade/staticpods.go

```diff
@@ -109,13 +112,100 @@ func (spm *KubeStaticPodPathManager) BackupManifestDir() string {
+	fmt.Println("[upgrade/staticpods] Waiting for the kubelet to restart the component")
+
+	// Wait for the mirror Pod hash to change; otherwise we'll run into race conditions here when the kubelet hasn't had time to
+	// notice the removal of the Static Pod, leading to a false positive below where we check that the API endpoint is healthy
+	// If we don't do this, there is a case where we remove the Static Pod manifest, kubelet is slow to react, kubeadm checks the
+	// API endpoint below of the OLD Static Pod component and proceeds quickly enough, which might lead to unexpected results.
+	if err := waiter.WaitForStaticPodControlPlaneHashChange(cfg.NodeName, component, beforePodHash); err != nil {
+		return rollbackOldManifests(recoverManifests, err, pathMgr)
+	}
+
+	// Wait for the static pod component to come up and register itself as a mirror pod
+	if err := waiter.WaitForPodsWithLabel("component=" + component); err != nil {
+		return rollbackOldManifests(recoverManifests, err, pathMgr)
+	}
+
+	fmt.Printf("[upgrade/staticpods] Component %q upgraded successfully!\n", component)
+	return nil
+}
 
-	// This string-string map stores the component name and backup filepath (if a rollback is needed).
-	// If a rollback is needed,
+func performEtcdStaticPodUpgrade(waiter apiclient.Waiter, pathMgr StaticPodPathManager, cfg *kubeadmapi.MasterConfiguration, recoverManifests map[string]string) error {
+	// Add etcd static pod spec only if external etcd is not configured
+	if len(cfg.Etcd.Endpoints) != 0 {
+		return fmt.Errorf("external etcd cannot be upgraded with kubeadm")
+	}
+
+	// Need to check currently used version and version from constants, if differs then uprgade
+	runningEtcdVersion, err := util.GetEtcdRunningVersion()
+	if err != nil {
+		return err
+	}
+	targetKubernetesVersion := version.MustParseGeneric(cfg.KubernetesVersion)
+	upgradeEtcdVersion, err := constants.GetEtcdSupportedVersion(uint8(targetKubernetesVersion.Minor()))
+	if err != nil {
+		return err
+	}
+	newEtcdVersion := version.MustParseGeneric(upgradeEtcdVersion)
+	if err != nil {
+		return err
+	}
+	// Ignore error here because the version struct was built with MustParse method
+	// which means any errors will trigger panic.
+	if res, _ := newEtcdVersion.Compare(runningEtcdVersion); res != 1 {
+		return fmt.Errorf("etcd upgrade was requested for the same or more recent version")
+	}
+
+	beforeEtcdPodHash, err := waiter.WaitForStaticPodSingleHash(cfg.NodeName, "etcd")
+	if err != nil {
+		return err
+	}
+
+	// Write the updated etcd static Pod manifest into the temporary directory
+	if err := etcdphase.CreateLocalEtcdStaticPodManifestFile(pathMgr.TempManifestDir(), cfg); err != nil {
+		return fmt.Errorf("error creating local etcd static pod manifest file: %v", err)
+	}
+
+	// Perform etcd upgrade using common to all control plane components function
+	return performStaticPodUpgrade("etcd", waiter, pathMgr, cfg, beforeEtcdPodHash, recoverManifests)
```

Comment:
nit: use the constant

Reply:
Please create a validateFunc for etcd here that checks version again, or the member health in the cluster

Reply:
So we have coverage that things are working